assert_writeln_magic.d

```diff
@@ -0,0 +1,312 @@
+                            fl[lineNr] = format("%swriteln(%s); // %s", indent, left, right);
+
+                        //writefln("line: %d, column: %d", lastAssert.line, lastAssert.column);
+                    }
+                }
+            }
+        }
+        resetTestState();
+    }
+
+private:
+
+    void resetTestState()
+    {
+        fromAssert = false;
+        fromStaticAssert = false;
+        lastEqualExpression = null;
+        lastAssert = null;
+    }
+
+    /// within in the node
+    bool inTest;
+    bool inAssert;
+    bool inFunctionCall;
+
+    /// at a sibling after the node was seen, but the upper parent hasn't been reached yet
+    bool fromAssert;
+    bool fromStaticAssert;
+
+    Rebindable!(const AssertExpression) lastAssert;
+    Rebindable!(const EqualExpression) lastEqualExpression;
+
+    string fileName;
+    FileLines fl;
+}
+
+void parseFile(string fileName)
+{
+    import dparse.lexer;
+    import dparse.parser : parseModule;
+    import dparse.rollback_allocator : RollbackAllocator;
+    import std.array : uninitializedArray;
+
+    auto inFile = File(fileName);
+    if (inFile.size == 0)
+        warningf("%s is empty", inFile.name);
+
+    ubyte[] sourceCode = uninitializedArray!(ubyte[])(to!size_t(inFile.size));
+    inFile.rawRead(sourceCode);
+    LexerConfig config;
+    auto cache = StringCache(StringCache.defaultBucketCount);
+    const(Token)[] tokens = getTokensForParser(sourceCode, config, &cache).array;
+
+    RollbackAllocator rba;
+    auto m = parseModule(tokens, fileName, &rba);
+    auto visitor = new TestVisitor(fileName);
+    visitor.visit(m);
+    delete visitor;
+}
+
+void main(string[] args)
+{
+    import std.file;
+    import std.getopt;
+    import std.path : asNormalizedPath;
+
+    string inputDir;
+    string[] ignoredFiles;
+
+    auto helpInfo = getopt(args, config.required,
+            "inputdir|i", "Folder to start the recursive search for unittest blocks (can be a single file)", &inputDir,
+            "ignore", "List of files to exclude (partial matching is supported)", &ignoredFiles);
+
+    if (helpInfo.helpWanted)
+    {
+        return defaultGetoptPrinter(`assert_writeln_magic
+Tries to lower EqualExpression in AssertExpressions of Unittest blocks to commented writeln calls.
+`, helpInfo.options);
+    }
+
+    inputDir = inputDir.asNormalizedPath.array;
+
+    DirEntry[] files;
+
+    if (inputDir.isFile)
+    {
+        files = [DirEntry(inputDir)];
+        inputDir = ".";
+    }
+    else
+    {
+        files = dirEntries(inputDir, SpanMode.depth).filter!(
+                a => a.name.endsWith(".d") && !a.name.canFind(".git")).array;
+    }
+
+    foreach (file; files)
+        if (!ignoredFiles.any!(x => file.name.canFind(x)))
+            file.name.parseFile;
+}
+
+/**
+A simple line-based in-memory representation of a file.
+ - will automatically write all changes when the objct is destructed
+ - will use a temporary file to do safe, whole file swaps
```

Comment:
BTW if you wanted safe, atomic file replacements I think writing to `<filename>.tmp` and renaming over it would be closer to that goal, since POSIX renames are atomic.